src/deco-theme.cpp

```diff
@@ -42,31 +44,95 @@ void decoration_theme_t::update_colors(void)
     bg_text = wf::color_t(bg_text_color);
 }
 
-/** @return The available height for displaying the title */
-int decoration_theme_t::get_font_height_px() const
+PangoFontDescription *create_font_description()
 {
-    static int font_sz = -1;
-    if (font_sz > 0)
+    GSettings *gs = g_settings_new("org.gnome.desktop.interface");
+
+    std::string title_font_val(title_font);
+    bool using_system_font{};
+
+    if (title_font_val.empty())
     {
-        return font_sz;
+        char *font = g_settings_get_string(gs, "font-name");
+        title_font_val = font;
+        using_system_font = true;
+        g_free(font);
+    }
+
+    PangoFontDescription *font_desc = pango_font_description_from_string(title_font_val.c_str());
+    int font_size = pango_font_description_get_size(font_desc);
+    bool font_size_absolute;
+
+    // font size will be 0 if nothing is specified - grab from system font if this is the case
+    // if the font is the system font, then just pray it works
+    if (!font_size && !using_system_font)
+    {
+        char *font = g_settings_get_string(gs, "font-name");
+        PangoFontDescription *sys_font_desc = pango_font_description_from_string(font);
+
+        font_size = pango_font_description_get_size(sys_font_desc);
+        font_size_absolute = pango_font_description_get_size_is_absolute(sys_font_desc);
+
+        pango_font_description_free(sys_font_desc);
+        g_free(font);
+    } else
+    {
+        font_size_absolute = pango_font_description_get_size_is_absolute(font_desc);
+    }
+
+    // scale the font size if we got it by text-scaling-factor
+    if (font_size)
+    {
+        double scaling_factor = g_settings_get_double(gs, "text-scaling-factor");
+        if (!scaling_factor) // should default to 1.0, but better safe than sorry
+        {
+            scaling_factor = 1.0;
+        }
+
+        if (font_size_absolute)
+        {
+            pango_font_description_set_absolute_size(font_desc, font_size * scaling_factor);
+        } else
+        {
+            pango_font_description_set_size(font_desc, font_size * scaling_factor);
+        }
     }
 
-    char *font = g_settings_get_string(gs, "font-name");
+    g_object_unref(gs);
+    return font_desc;
+}
+
+PangoFontDescription *get_font_description()
+{
+    static std::unique_ptr<PangoFontDescription, decltype(&pango_font_description_free)> font_desc(
+        create_font_description(), &pango_font_description_free);
+
+    static std::once_flag once_flag;
+    std::call_once(once_flag, [] {
+        title_font.set_callback([] {
+            font_desc.reset(create_font_description());
```

Comment:
If I am reading this correctly, doesn't the `font_desc` pointer returned by `create_font_description()` need to be freed somewhere somehow?

Reply:
The pointer returned by create_font_description is managed by font_desc in this case, which is static and thus is only created once. This management means calling pango_font_description_free when font_desc ceases to exist which, due to it being static, means when the calling program (Wayfire) exits.

get_font_description returns the pointer which font_desc manages. Because it's managed, it doesn't have to be manually freed.

All of this means the calculation of the font description, in addition to all of the modifications done to it, only have to be done once ever (unless if the title font option is changed, in which it's done again), and the font description instance is properly cleaned up when it's no longer needed.

I've made an example showing this idea in action [here](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEIJIAzKQADqgKhE4MHt6%2B/ilpGQIhYZEsMXGJtpj2jgJCBEzEBNk%2BfgF2mA6Z9Y0ExRHRsfFJCg1NLbntY32hA2VDiQCUtqhexMjsHOYJocjeWADUJgluTqPEmKzH2CYaAILbu/uYRydsLCQAnte3D/fnXgcByiBhBTDBAIcvxMAHYrPcDoiDqN0CAQOdQsADugmA1jvCHjCACL46H3MEUgyQggAKmBoIZTAA%2BsgLrjMBAUWiMUYmQA3PCYADu2NxTEW0LhvyRyIIqJQqwIrzcxzcRzMZkp4IMLLZBA5Eo1ytVsvlrlopIRSK1EIIxEBtIOiteRIOYRFNqpdodlruMsVAFprjiGi7RXiEgSZRcCGsGE6vARfbCSfdfnzUHh0PTtbmmVQLhzPUxqXTQqMJWmpVbEVyFYnjSd1ZrGWD84WIIazI21XXzb6ZVh6PqDuWk5HJam/ndi9SDjS0AxRgdgJgCEzi53JVGkXW0A3VSbzGZV%2BvN12e6a0f2JzXZbi8MgryAvAw8ABHLyYJnJO2q2feg4pDYp0tAEJ8yQcuYABsxbtpgmASgk2AHK%2BH5fhA0oytarY6qylz6phGoaMeyw5lq8GIQOSIxnGqFvp%2BmAAHSnlut6ElO6aZtmuwXGwggbrhTBblWO6IgB9oOHSxZhqegm5mCbFiaOggHE%2BxyunWoyZhAxZBshIbitR4lCfpuBimGdZEEyPLAERViWMgSEEimZJ3KESosEwoQiYSyn8MQBwQB5o5hiRkahSaXDhRY6oORYeCVn62GIrxmD8WeQlKVhtZymi%2B5KoeTZyeeZmGZefYMOgFrsSmHDLLQnAAKy8H4HBaKQqCcCqljWMiqzrC82w8KQBCaPVywANYgAkCRMbNC2LUt0H6JwkiteNnWcLwCggBoo3jcscCwEgaAsMkdCxOQlBnRd9BxMAUgBDQYGxLtumbVEoSNJ8nAjV9zDEJ8ADyUTaJ0Y3cLwZ0ZcDDC0L97W8FgUReMAbhiLQu1Q6QWBeUY4hI7jeAXF0fKYNjHWYKonSJpsHUedUm20HgUTED9HhYJtdp4Cwf28OTxBRGkmBEulhjACzRiHXwBjAAoABqgpCsDkFtSN/CCCIYjsFIMiCIoKjqETuhcPoEsoNY1j6Kzu2QMsqC/pk2O8KggvEFmFPwMsHRdM4ECuBMfhm8EsylOUeipOktRZJ4rSRwUMf9OHQxm77Mc9OMce5Gn1QQ900zJ4McRp9MQd6KMvRF/MJc%2BwNGwSA1zUbUTXUcAcqgABzQQG0GSCuyBPlITHdhAuCECQ6oJFwiy8JDWiLFNM1zUtq8LStjUcOtpBtR1bc7XtB1I0dMCICAirJIm10QLdl3EOErCbF3Pd9wPQ%2BSCPvCYPgRAe6iZua2EKIcQetAGGzUJtU2pAhTs2SPzJuHAWo702m3YGiZL5KlQFQDu3de792AIPA4w9R4eHOnfKeM856HSXrNeaa9V6rS3i3Pe21bCH3nhNRhZhmGu1YRwxepBBbpGcJIIAA%3D%3D).

Perhaps I could make another commit adding a comment which shortly describes this, and making get_font_description return a const pointer so it can't be manually freed.

Reply:
You also have to consider pixdecor can be reloaded since it's a wayfire plugin. Also we don't want to leak on settings change. Not sure if these cases are problematic but I'd prefer explicit cleanup if possible.

Reply:
Did some debugging to be sure and there are no leaks on plugin reload or settings change. When those happen, the new font description is created, the old one is deleted, and it takes the old one's place in being managed. Furthermore, I have used this plugin with this code for many hours long sessions for the past few days with no issues.

I did have a concern with the explicit cleanup route and it's why I went this route in the first place - as far as I know, a new font description would have to be created every time, and the font description is grabbed TONS of times with normal usage. This may have been fine previously (and who knows, may still be), but I was afraid the additional computations with the font option and scaling and such might create significant overhead in some cases. This route avoids that and is the safest way to go about doing so.

Reply:
Thanks for having a closer look and for your detailed explanation! I will test this when I have time and make sure things work as expected. Thanks for the pull request, and your interest in wayfire and pixdecor.